extensions/standard-processors/processors/RetryFlowFile.cpp

```diff
@@ -0,0 +1,212 @@
+}
+
+void RetryFlowFile::readDynamicPropertyKeys(core::ProcessContext* context) {
+  exceeded_flowfile_attribute_keys.clear();
+  const std::vector<std::string> dynamic_prop_keys = context->getDynamicPropertyKeys();
+  logger_->log_info("RetryFlowFile registering %d keys", dynamic_prop_keys.size());
+  for (const auto& key : dynamic_prop_keys) {
+    exceeded_flowfile_attribute_keys.emplace_back(core::PropertyBuilder::createProperty(key)->withDescription("auto generated")->supportsExpressionLanguage(true)->build());
+    logger_->log_info("RetryFlowFile registered attribute '%s'", key);
+  }
+}
+
+// Returns (1, true) on non-numerical or out-of-bounds retry value
+std::pair<uint64_t, bool> RetryFlowFile::getRetryPropertyValue(const std::shared_ptr<FlowFileRecord>& flow_file) {
+  std::string value_as_string;
+  try {
+    if (flow_file->getAttribute(retry_attribute_, value_as_string)) {
+      return std::make_pair(std::stoul(value_as_string), false);
+    }
+  }
+  catch(const std::invalid_argument&) {
+    if (fail_on_non_numerical_overwrite_) {
+      logger_->log_info("Non-numerical retry property in RetryFlowFile. Sending flowfile to failure...", value_as_string);
+      return std::make_pair(1, true);
+    }
+    logger_->log_info("Non-numerical retry property in RetryFlowFile: overwriting %s with 1.", value_as_string);
+  }
+  catch(const std::out_of_range&) {
+    logger_->log_error("Narrowing Exception for %s, treating it as non-numerical value", value_as_string);
+  }
+  return std::make_pair(1, false);
+}
+
+// Returns true on fail on reuse scenario
+bool RetryFlowFile::updateUUIDMarkerAndCheckFailOnReuse(const std::shared_ptr<FlowFileRecord>& flow_file) {
+  const std::string last_retried_by_property_name = retry_attribute_ + ".uuid";
+  const std::string current_processor_uuid = getUUIDStr();
+  std::string last_retried_by_uuid;
+  if (flow_file->getAttribute(last_retried_by_property_name, last_retried_by_uuid)) {
+    if (last_retried_by_uuid != current_processor_uuid) {
+      if (reuse_mode_ == FAIL_ON_REUSE) {
+        logger_->log_error("FlowFile %s was previously retried with the same attribute by a different "
+            "processor (uuid: %s, current uuid: %s). Transfering flowfile to 'failure'...",
+            flow_file->getUUIDStr(), last_retried_by_uuid, current_processor_uuid);
+        return true;
+      }
+      if (reuse_mode_ == WARN_ON_REUSE) {
+        logger_->log_warn("Reusing retry attribute that belongs to different processor. Resetting value to 1.");
+      } else {  // Assuming reuse_mode_ == RESET_REUSE
+        logger_->log_debug("Reusing retry attribute that belongs to different processor. Resetting value to 1.");
+      }
+    }
+  }
+  flow_file->setAttribute(last_retried_by_property_name, getUUIDStr());
+  return false;
+}
+
+bool RetryFlowFile::setRetriesExceededAttributesOnFlowFile(core::ProcessContext* context, const std::shared_ptr<FlowFileRecord>& flow_file) {
```

Comment:
Const!

I don't think this function should return true or false. 

This should be void. In case something throws when trying to set attributes (shouldn't), just let the exception fly and that results in a rollback.

Reply:
Will be marked const.

Reply:
Everything done in the given onTrigger execution (adding/modifying attributes, routing, creating flowfiles) is rolled back.

Reply:
Will change this to throw.